[SPARK-14867][BUILD] Remove `--force` option in `build/mvn`

dongjoon-hyun opened the pull request:
## What changes were proposed in this pull request?

Currently, `build/mvn` provides a convenient option, `--force`, in order to use the recommended version of maven without changing PATH environment variable. However, there were two problems.
- `dev/lint-java` does not use the newly installed maven.
  
  ``` bash
  $ ./build/mvn --force clean
  $ ./dev/lint-java 
  Using `mvn` from path: /usr/local/bin/mvn
  ```
- It's not easy to type `--force` option always.

If '--force' option is used once, we had better prefer the installed maven recommended by Spark.
This PR makes `build/mvn` check the existence of maven installed by `--force` option first.

According to the comments, this PR aims to the followings:
- Detect the maven version from `pom.xml`.
- Install maven if there is no or old maven.
- Remove `--force` option.
## How was this patch tested?

Manual.

``` bash
$ ./build/mvn --force clean
$ ./dev/lint-java 
Using `mvn` from path: /Users/dongjoon/spark/build/apache-maven-3.3.9/bin/mvn
...
$ rm -rf ./build/apache-maven-3.3.9/
$ ./dev/lint-java 
Using `mvn` from path: /usr/local/bin/mvn
```

dongjoon-hyun committed:
[SPARK-14867][BUILD] Make `build/mvn` to use the downloaded maven if it exist.

SparkQA commented:
**[Test build #56780 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56780/consoleFull)** for PR 12631 at commit [`56355fa`](https://github.com/apache/spark/commit/56355fab23ef59b79bbaafa7164bbbb3ca742326).

SparkQA commented:
**[Test build #56780 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56780/consoleFull)** for PR 12631 at commit [`56355fa`](https://github.com/apache/spark/commit/56355fab23ef59b79bbaafa7164bbbb3ca742326).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56780/
Test PASSed.

srowen commented:
Yes, on a system with Maven < 3.3.9 installed, you always have to pass `--force`. This would at least make it optional on the second or subsequent run. 

Ideally we could even auto-detect the `mvn` version, which isn't too hard, and then download and/or use the downloaded copy only if the installed one is too old. What do you think about trying that? I wouldn't want to do it if it ends up being a lot of complexity, but ought to be a few more lines. We could even remove `--force` then, I suppose, since you have to download it if your local maven is too old.

dongjoon-hyun commented:
That sounds good to me. No problem!
I'll update today. 
Thank you, @srowen !

dongjoon-hyun committed:
Choose Maven based on versions.

SparkQA commented:
**[Test build #56808 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56808/consoleFull)** for PR 12631 at commit [`afd7f03`](https://github.com/apache/spark/commit/afd7f0386c3d467b4bc22e9ae15e77dd279149e7).

SparkQA commented:
**[Test build #56808 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56808/consoleFull)** for PR 12631 at commit [`afd7f03`](https://github.com/apache/spark/commit/afd7f0386c3d467b4bc22e9ae15e77dd279149e7).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56808/
Test FAILed.

dongjoon-hyun committed:
Change to local variables.

SparkQA commented:
**[Test build #56809 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56809/consoleFull)** for PR 12631 at commit [`ee883e0`](https://github.com/apache/spark/commit/ee883e042de154282df75ff2c9a3c818dd431240).

SparkQA commented:
**[Test build #56809 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56809/consoleFull)** for PR 12631 at commit [`ee883e0`](https://github.com/apache/spark/commit/ee883e042de154282df75ff2c9a3c818dd431240).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56809/
Test FAILed.

dongjoon-hyun committed:
Add a code for backward compatability.

SparkQA commented:
**[Test build #56810 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56810/consoleFull)** for PR 12631 at commit [`d593607`](https://github.com/apache/spark/commit/d593607c8628db7b485cbcde5c2f6746fb96b73e).

markhamstra commented on build/mvn:
```diff
@@ -69,14 +69,23 @@ install_app() {
 # Install maven under the build/ folder
 install_mvn() {
-  local APACHE_MIRROR=${APACHE_MIRROR:-'https://www.apache.org/dyn/closer.lua?action=download&filename='}
+  MVN_VERSION="3.3.9"
```
This really should be grep'd from the pom file, as is done with scala_version in install_scala.  Also, make sure that the correct maven in downloaded and installed in the case that there isn't any mvn available when this script is run.

dongjoon-hyun replied:
Thank you, @markhamstra .
I'll fix that to read "3.3.9" from pom file.

By the way, what do you mean by "make sure that the correct maven is downloaded and installed in the case that there isn't any mvn available when this script is run"?

markhamstra replied:
I mean that using this build/mvn script should work even if there isn't any version of maven currently installed on the machine (which might be the case, e.g., when using an automated tool to set up a new build node.)  I believe that would work with `--force` previously, and it should work when there isn't a `--force` option anymore.

dongjoon-hyun replied:
Ah. For that, I tested already. When there is no maven, this script installs the correct maven. Thank you again!

SparkQA commented:
**[Test build #56810 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56810/consoleFull)** for PR 12631 at commit [`d593607`](https://github.com/apache/spark/commit/d593607c8628db7b485cbcde5c2f6746fb96b73e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56810/
Test PASSed.

dongjoon-hyun committed:
Determine the maven version from `pom.xml`.

SparkQA commented:
**[Test build #56829 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56829/consoleFull)** for PR 12631 at commit [`cfe9b4e`](https://github.com/apache/spark/commit/cfe9b4eb101b9497a76938967b931ba99bbe6c9f).

dongjoon-hyun commented:
Rebased.

SparkQA commented:
**[Test build #56830 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56830/consoleFull)** for PR 12631 at commit [`8504cce`](https://github.com/apache/spark/commit/8504cce65231c38e70278c4c625aa3c87cb4c0c3).

SparkQA commented:
**[Test build #56829 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56829/consoleFull)** for PR 12631 at commit [`cfe9b4e`](https://github.com/apache/spark/commit/cfe9b4eb101b9497a76938967b931ba99bbe6c9f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56829/
Test PASSed.

SparkQA commented:
**[Test build #56830 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56830/consoleFull)** for PR 12631 at commit [`8504cce`](https://github.com/apache/spark/commit/8504cce65231c38e70278c4c625aa3c87cb4c0c3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56830/
Test FAILed.

SparkQA commented:
**[Test build #56836 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56836/consoleFull)** for PR 12631 at commit [`b11de8a`](https://github.com/apache/spark/commit/b11de8af931d0390d4d2e89aba61791410fe6984).

SparkQA commented:
**[Test build #56836 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56836/consoleFull)** for PR 12631 at commit [`b11de8a`](https://github.com/apache/spark/commit/b11de8af931d0390d4d2e89aba61791410fe6984).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56836/
Test FAILed.

srowen commented on build/mvn:
```diff
@@ -67,17 +67,27 @@ install_app() {
-  local MVN_VERSION="3.3.9"
-  local APACHE_MIRROR=${APACHE_MIRROR:-'https://www.apache.org/dyn/closer.lua?action=download&filename='}
+  local MVN_VERSION=`grep "<maven.version>" "${_DIR}/../pom.xml" | \
+                     head -1 | cut -f2 -d'>' | cut -f1 -d'<'`
```
Since you use `awk` below, would this find the version correctly? `awk -F '[<>]' '{print $3}'` Looks like it, but I don't know if it's much simpler. You might also use "-n1" consistently like below in the `head` command.

dongjoon-hyun replied:
It works as the Jenkins runs.
```
spark:SPARK-14867$ grep "<maven.version>" pom.xml 
    <maven.version>3.3.9</maven.version>
```

For simplification, I'm not sure since I borrowed it from the following `install_scala()` code in the same script. Should I change them both for consistency?
```bash
install_scala() {
  # determine the Scala version used in Spark
  local scala_version=`grep "scala.version" "${_DIR}/../pom.xml" | \
                       head -1 | cut -f2 -d'>' | cut -f1 -d'<'`
```
Please let me know. If needed, I will update both.

srowen commented on build/mvn:
```diff
@@ -116,23 +126,15 @@ install_scala() {
+# Remove `--force` for backward compatibility.
 if [ "$1" == "--force" ]; then
-  FORCE_MVN=1
   shift
```
You could emit an info message for `--force`; no big deal

dongjoon-hyun replied:
Sure! Done

srowen commented on build/mvn:
```diff
@@ -67,17 +67,27 @@ install_app() {
+  fi
+  # See simple version normalization: http://stackoverflow.com/questions/16989598/bash-comparing-version-numbers
+  function version { echo "$@" | awk -F. '{ printf("%03d%03d%03d\n", $1,$2,$3); }'; }
+  if [ $(version $MVN_DETECTED_VERSION) -lt $(version $MVN_VERSION) ]; then
```
If there's no local version, then this always returns "false" as if the detected version is 0?

dongjoon-hyun replied:
When there is no local version, it return true.

dongjoon-hyun committed:
Add a deprecation message for option --force

SparkQA commented:
**[Test build #56855 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56855/consoleFull)** for PR 12631 at commit [`128618e`](https://github.com/apache/spark/commit/128618e3497f2698892ff50a3c5f0fa84c8ff936).

SparkQA commented:
**[Test build #56855 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56855/consoleFull)** for PR 12631 at commit [`128618e`](https://github.com/apache/spark/commit/128618e3497f2698892ff50a3c5f0fa84c8ff936).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56855/
Test FAILed.

SparkQA commented:
**[Test build #56857 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56857/consoleFull)** for PR 12631 at commit [`2b54019`](https://github.com/apache/spark/commit/2b54019ccbe171aae0c6f05ea71ad11aa1627dd7).

SparkQA commented:
**[Test build #56857 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56857/consoleFull)** for PR 12631 at commit [`2b54019`](https://github.com/apache/spark/commit/2b54019ccbe171aae0c6f05ea71ad11aa1627dd7).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56857/
Test FAILed.

SparkQA commented:
**[Test build #56859 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56859/consoleFull)** for PR 12631 at commit [`497dcb5`](https://github.com/apache/spark/commit/497dcb535b25031aa56db90edd7c71e84ae8c62e).

SparkQA commented:
**[Test build #56859 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56859/consoleFull)** for PR 12631 at commit [`497dcb5`](https://github.com/apache/spark/commit/497dcb535b25031aa56db90edd7c71e84ae8c62e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56859/
Test FAILed.

dongjoon-hyun commented:
The test failure seems to be irrelevant to this PR.
```
[info] Exception encountered when attempting to run a suite with class name: org.apache.spark.streaming.MapWithStateSuite *** ABORTED *** (41 seconds, 299 milliseconds)
[info]   java.io.IOException: Failed to delete: /home/jenkins/workspace/SparkPullRequestBuilder/streaming/checkpoint/spark-d2368db9-6d2a-43ba-b79f-372e51a41ce0
```

dongjoon-hyun committed:
Update test-dependecies.sh.

dongjoon-hyun commented:
I updated the commit time of the last commit to retrigger Jenkins.

SparkQA commented:
**[Test build #56865 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56865/consoleFull)** for PR 12631 at commit [`55e4c5b`](https://github.com/apache/spark/commit/55e4c5be0b7d5a893af8dd898d5125a02cad41c3).

SparkQA commented:
**[Test build #56865 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56865/consoleFull)** for PR 12631 at commit [`55e4c5b`](https://github.com/apache/spark/commit/55e4c5be0b7d5a893af8dd898d5125a02cad41c3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56865/
Test PASSed.

dongjoon-hyun committed:
Use awk command according to the comments.

SparkQA commented:
**[Test build #56890 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56890/consoleFull)** for PR 12631 at commit [`20a520a`](https://github.com/apache/spark/commit/20a520a35708d6ad6077c78e5e021e4eddbb60a3).

dongjoon-hyun commented:
Hi, @srowen .
Now, it's ready for review again.
I think I've addressed all your comments. The last one was about `awk` command that you recommended.
Currently, Jenkins passes the followings.
```
========================================================================
Running build tests
========================================================================
exec: curl -s -L https://downloads.typesafe.com/zinc/0.3.9/zinc-0.3.9.tgz
exec: curl -s -L https://downloads.typesafe.com/scala/2.11.8/scala-2.11.8.tgz
exec: curl -s -L https://www.apache.org/dyn/closer.lua?action=download&filename=/maven/maven-3/3.3.9/binaries/apache-maven-3.3.9-bin.tar.gz
...
```
Thank you for your review always. If there is anything to do more, please let me know.

srowen commented:
Yeah that looks good to me.

dongjoon-hyun commented:
Thank you, @srowen !

SparkQA commented:
**[Test build #56890 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56890/consoleFull)** for PR 12631 at commit [`20a520a`](https://github.com/apache/spark/commit/20a520a35708d6ad6077c78e5e021e4eddbb60a3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56890/
Test PASSed.

dongjoon-hyun commented:
Hi, @srowen .
Could you merge this PR?

srowen commented:
Merged to master

dongjoon-hyun commented:
Thank you, @srowen !
